### PR TITLE
Additional explanation of middleware pipeline configuration options.

### LINKED ISSUE #15497: Use, Run, and Map - differentiate them
[Use, Run, and Map](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-3.0#use-run-and-map)

Does a poor job differentiating them.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 9898482c-fa54-7ce2-1716-d14fbe03990d
* Version Independent ID: 3926c686-0ccc-f9fe-ca52-0134612a1623
* Content: [ASP.NET Core Middleware](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-3.0#use-run-and-map)
* Content Source: [aspnetcore/fundamentals/middleware/index.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/fundamentals/middleware/index.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-fundamentals**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by serpent5**
Fixes #15497.

### CODE REVIEW

File: aspnetcore/fundamentals/middleware/index.md [hunk 1]
```diff
@@ -171,17 +181,6 @@ The following table shows the requests and responses from `http://localhost:1234
...
-
-[!code-csharp[](index/snapshot/Chain/StartupMapWhen.cs)]
-
-The following table shows the requests and responses from `http://localhost:1234` using the previous code.
-
-| Request                       | Response                     |
-| ----------------------------- | ---------------------------- |
-| localhost:1234                | Hello from non-Map delegate. |
-| localhost:1234/?branch=master | Branch used = master         |
-
```
Comment (serpent5) [inline aspnetcore/fundamentals/middleware/index.md]: I moved this down. There was a bit of jumping between `Map`, `MapWhen`, and then `Map` again.

File: aspnetcore/fundamentals/middleware/index.md [hunk 2]
```diff
@@ -37,11 +37,9 @@ The simplest possible ASP.NET Core app sets up a single request delegate that ha
 
 [!code-csharp[](index/snapshot/Middleware/Startup.cs)]
 
-The first <xref:Microsoft.AspNetCore.Builder.RunExtensions.Run*> delegate terminates the pipeline.
```
Comment (serpent5) [inline aspnetcore/fundamentals/middleware/index.md]: I moved this down to a more detailed comparison between `Use` and `Run`.

File: aspnetcore/fundamentals/middleware/index.md [hunk 3]
```diff
@@ -152,9 +164,7 @@ public void Configure(IApplicationBuilder app)
 }
 ```
 
-## Use, Run, and Map
-
-Configure the HTTP pipeline using <xref:Microsoft.AspNetCore.Builder.UseExtensions.Use*>, <xref:Microsoft.AspNetCore.Builder.RunExtensions.Run*>, and <xref:Microsoft.AspNetCore.Builder.MapExtensions.Map*>. The `Use` method can short-circuit the pipeline (that is, if it doesn't call a `next` request delegate). `Run` is a convention, and some middleware components may expose `Run[Middleware]` methods that run at the end of the pipeline.
```
Comment (serpent5) [inline aspnetcore/fundamentals/middleware/index.md]: The first two lines here are already covered. I moved the last line to the section where I explain more about `Run`.

File: aspnetcore/fundamentals/middleware/index/snapshot/Chain/StartupUseWhen.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
+﻿public class Startup
+{
+    private readonly ILogger<Startup> _logger;
+
+    public Startup(ILogger<Startup> logger)
```
Comment (serpent5) [inline aspnetcore/fundamentals/middleware/index/snapshot/Chain/StartupUseWhen.cs]: I'm taking advantage of the fact that this is 2.2 and injecting a logger into the constructor.

File: aspnetcore/fundamentals/middleware/index/snapshot/Chain/StartupUseWhen.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    private void HandleBranchAndRejoin(IApplicationBuilder app)
+    {
+        app.Use(async (context, next) =>
+        {
+            var branchVer = context.Request.Query["branch"];
+            _logger.LogInformation("Branch used = {branchVer}", branchVer);
+
+            await next();
```
Comment (serpent5) [inline aspnetcore/fundamentals/middleware/index/snapshot/Chain/StartupUseWhen.cs]: Should I add a comment here to be explicit about why we're logging and no longer trying to write to the response? It's already explained earlier in the topic, but might be useful here too. i.e.:

```csharp
// Do work that doesn't write to the Response.
await next();
```

**Comment by Rick-Anderson** [comment]
Looks good
External review:

* [Download 1.zip](https://www.dropbox.com/sh/yy994551cje0kap/AAB0tRnm7gze2kYrSYPSFHIda?dl=0)
* Unzip 1.zip file

**Comment by serpent5** [comment]
Thanks. I _think_ that all looks OK.

File: aspnetcore/fundamentals/middleware/index.md [hunk 6]
```diff
@@ -199,6 +190,23 @@ app.Map("/level1", level1App => {
...
+| Request                       | Response                     |
+| ----------------------------- | ---------------------------- |
+| localhost:1234                | Hello from non-Map delegate. |
+| localhost:1234/?branch=master | Branch used = master         |
+
+<xref:Microsoft.AspNetCore.Builder.UseWhenExtensions.UseWhen*> also branches the request pipeline based on the result of the given predicate. Unlike with `MapWhen`, this branch is rejoined to the main pipeline if it does short-circuit or contain a terminal middleware:
+
+[!code-csharp[](index/snapshot/Chain/StartupUseWhen.cs?highlight=23-24)]
+
+In the preceding example, a response of "Hello from main pipeline." is written for all requests. If the request includes a query string variable `branch`, its value is logged before the main pipeline is rejoined.
```
Comment (Rick-Anderson) [inline aspnetcore/fundamentals/middleware/index.md]: I shortened this up.

Comment (serpent5) [reply aspnetcore/fundamentals/middleware/index.md]: Yeah, that's definitely better. I overdid it with the short sentence thing. Thanks.

**Pull request merged**